ee/query-service/rules/anomaly.go

```diff
@@ -0,0 +1,392 @@
+) (*AnomalyRule, error) {
+
+	zap.L().Info("creating new AnomalyRule", zap.String("id", id), zap.Any("opts", opts))
+
+	baseRule, err := baserules.NewBaseRule(id, p, reader, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	t := AnomalyRule{
+		BaseRule: baseRule,
+	}
+
+	if strings.ToLower(p.RuleCondition.Seasonality) == "hourly" {
+		t.seasonality = anomaly.SeasonalityHourly
+	} else if strings.ToLower(p.RuleCondition.Seasonality) == "daily" {
+		t.seasonality = anomaly.SeasonalityDaily
+	} else if strings.ToLower(p.RuleCondition.Seasonality) == "weekly" {
+		t.seasonality = anomaly.SeasonalityWeekly
+	} else {
+		t.seasonality = anomaly.SeasonalityDaily
+	}
+
+	zap.L().Info("using seasonality", zap.String("seasonality", t.seasonality.String()))
+
+	querierOptsV2 := querierV2.QuerierOptions{
+		Reader:        reader,
+		Cache:         cache,
+		KeyGenerator:  queryBuilder.NewKeyGenerator(),
+		FeatureLookup: featureFlags,
+	}
+
+	t.querierV2 = querierV2.NewQuerier(querierOptsV2)
+	t.reader = reader
+	if t.seasonality == anomaly.SeasonalityHourly {
+		t.provider = anomaly.NewHourlyProvider(
+			anomaly.WithCache[*anomaly.HourlyProvider](cache),
+			anomaly.WithKeyGenerator[*anomaly.HourlyProvider](queryBuilder.NewKeyGenerator()),
+			anomaly.WithReader[*anomaly.HourlyProvider](reader),
+			anomaly.WithFeatureLookup[*anomaly.HourlyProvider](featureFlags),
+		)
+	} else if t.seasonality == anomaly.SeasonalityDaily {
+		t.provider = anomaly.NewDailyProvider(
+			anomaly.WithCache[*anomaly.DailyProvider](cache),
+			anomaly.WithKeyGenerator[*anomaly.DailyProvider](queryBuilder.NewKeyGenerator()),
+			anomaly.WithReader[*anomaly.DailyProvider](reader),
+			anomaly.WithFeatureLookup[*anomaly.DailyProvider](featureFlags),
+		)
+	} else if t.seasonality == anomaly.SeasonalityWeekly {
+		t.provider = anomaly.NewWeeklyProvider(
+			anomaly.WithCache[*anomaly.WeeklyProvider](cache),
+			anomaly.WithKeyGenerator[*anomaly.WeeklyProvider](queryBuilder.NewKeyGenerator()),
+			anomaly.WithReader[*anomaly.WeeklyProvider](reader),
+			anomaly.WithFeatureLookup[*anomaly.WeeklyProvider](featureFlags),
+		)
+	}
+	return &t, nil
+}
+
+func (r *AnomalyRule) Type() baserules.RuleType {
+	return RuleTypeAnomaly
+}
+
+func (r *AnomalyRule) prepareQueryRange(ts time.Time) (*v3.QueryRangeParamsV3, error) {
+
+	zap.L().Info("prepareQueryRange", zap.Int64("ts", ts.UnixMilli()), zap.Int64("evalWindow", r.EvalWindow().Milliseconds()), zap.Int64("evalDelay", r.EvalDelay().Milliseconds()))
+
+	start := ts.Add(-time.Duration(r.EvalWindow())).UnixMilli()
+	end := ts.UnixMilli()
+
+	if r.EvalDelay() > 0 {
+		start = start - int64(r.EvalDelay().Milliseconds())
+		end = end - int64(r.EvalDelay().Milliseconds())
+	}
+	// round to minute otherwise we could potentially miss data
+	start = start - (start % (60 * 1000))
+	end = end - (end % (60 * 1000))
+
+	compositeQuery := r.Condition().CompositeQuery
+
+	if compositeQuery.PanelType != v3.PanelTypeGraph {
```

Comment:
All alerts are "graph" because we fetch the time series data for the eval duration and use the condition settings. The value type panel is also a time series that reduces the values to a single value. The equivalent reduction can be performed using alert conditions.

reductTo:
1. sum -> InTotal
2. average -> OnAverage
3. min -> Below + all the times
4. max -> above + all the times